src/USER-BROWNIAN/fix_brownian_sphere.cpp

```diff
@@ -0,0 +1,321 @@
+
+enum{NONE,DIPOLE};
+
+/* ---------------------------------------------------------------------- */
+
+FixBrownianSphere::FixBrownianSphere(LAMMPS *lmp, int narg, char **arg) :
+  Fix(lmp, narg, arg)
+{
+  virial_flag = 1;
+
+  time_integrate = 1;
+
+  extra = NONE;
+
+  if (narg > 11 || narg < 8 )
+    error->all(FLERR,"Illegal fix brownian/sphere command.");
+
+  if (!atom->sphere_flag)
+    error->all(FLERR,"Fix brownian/sphere requires atom style sphere");
+
+  gamma_t = utils::numeric(FLERR,arg[3],false,lmp);
+  if (gamma_t <= 0.0)
+    error->all(FLERR,"Fix brownian/sphere translational viscous drag "
+	       "coefficient must be > 0.");
+
+  gamma_r = utils::numeric(FLERR,arg[4],false,lmp);
+  if (gamma_t <= 0.0)
+    error->all(FLERR,"Fix brownian/sphere rotational viscous drag "
+	       "coefficient must be > 0.");
+
+
+  diff_t = utils::numeric(FLERR,arg[5],false,lmp);
+  if (diff_t <= 0.0)
+    error->all(FLERR,"Fix brownian/sphere translational diffusion "
+	       "coefficient must be > 0.");
+  
+  diff_r = utils::numeric(FLERR,arg[6],false,lmp);
+  if (diff_r <= 0.0)
+    error->all(FLERR,"Fix brownian/sphere rotational diffusion "
+	       "coefficient must be > 0.");
+  
+  seed = utils::inumeric(FLERR,arg[7],false,lmp);
+  if (seed <= 0) error->all(FLERR,"Fix brownian/sphere seed must be > 0.");
+
+  noise_flag = 1;
+  gaussian_noise_flag = 0;
+
+  int iarg = 8;
+
+  while (iarg < narg) {
+    if (strcmp(arg[iarg],"rng") == 0) {
+      if (narg == iarg + 1) {
+	error->all(FLERR,"Illegal fix/brownian/sphere command.");
+      }
+      if (strcmp(arg[iarg + 1],"uniform") == 0) {
+	noise_flag = 1;
+      } else if (strcmp(arg[iarg + 1],"gaussian") == 0) {
+	noise_flag = 1;
+	gaussian_noise_flag = 1;
+      } else if (strcmp(arg[iarg + 1],"none") == 0) {
+	noise_flag = 0;
+      } else {
+	error->all(FLERR,"Illegal fix/brownian/sphere command.");
+      }
+      iarg = iarg + 2;
+    } else if (strcmp(arg[iarg],"dipole") == 0) {
+      extra = DIPOLE;
+      iarg = iarg + 1;
+    } else {
+      error->all(FLERR,"Illegal fix/brownian/sphere command.");
+    }
+  }
+  
+  if (extra == DIPOLE && !atom->mu_flag)
+    error->all(FLERR,"Fix brownian/sphere update dipole requires atom attribute mu");
+
+  // initialize Marsaglia RNG with processor-unique seed
+  random = new RanMars(lmp,seed + comm->me);
+
+}
+
+/* ---------------------------------------------------------------------- */
+
+int FixBrownianSphere::setmask()
+{
+  int mask = 0;
+  mask |= INITIAL_INTEGRATE;
+  mask |= POST_FORCE;
+  return mask;
+}
+
+/* ---------------------------------------------------------------------- */
+
+FixBrownianSphere::~FixBrownianSphere()
+{
+  delete random;
+}
+
+
+
+/* ---------------------------------------------------------------------- */
+
+void FixBrownianSphere::init()
+{
+  
+  g1 =  force->ftm2v/gamma_t;
+  g3 = force->ftm2v/gamma_r;
+  if (noise_flag == 0) {
+    g2 = 0;
+    g4 = 0;
+    rng_func = &RanMars::zero_rng;
```

Comment:
rather than using function pointers (and adding convenience functions to the pRNG class), I would suggest to just use if statements here and in the post_force() code. To avoid performance penalties, you can follow the example of fix langevin, which introduces a function template for different modes (and combination of modes) which are then selected in post_force() so the actual "if" is done outside the inner loop(s) and the compiler will then happily optimize away the if statements in the templated functions and effectively create multiple instances of the template as needed at compile time without requiring to program them explicitly.

Reply:
Done (I think), I'm still somewhat new to C++ so I'm not super confident in this, but everything is working as I would expect (e.g. diffusion in non-interacting systems is correct).